pkg/provision/metadata/metadata.go

```diff
@@ -0,0 +1,152 @@
+	// resolved plugins and parent) DevWorkspace yaml
+	flattenedYamlFilename = "flattened.devworkspace.yaml"
+
+	// metadataMountPath is where files containing workspace metadata are mounted
+	metadataMountPath = "/devworkspace-metadata"
+)
+
+// ProvisionWorkspaceMetadata creates a configmap on the cluster that stores metadata about the workspace and configures all
+// workspace containers to mount that configmap at /devworkspace-metadata. Each container has the environment
+// variable DEVWORKSPACE_METADATA set to the mount path for the configmap
+func ProvisionWorkspaceMetadata(podAdditions *v1alpha1.PodAdditions, original, flattened *dw.DevWorkspace, api *provision.ClusterAPI) error {
+	cm, err := getSpecMetadataConfigMap(original, flattened)
+	if err != nil {
+		return err
+	}
+	err = controllerutil.SetControllerReference(original, cm, api.Scheme)
+	if err != nil {
+		return err
+	}
+	if inSync, err := syncConfigMapToCluster(cm, api); err != nil {
+		return err
+	} else if !inSync {
+		return &NotReadyError{
+			Message: "Waiting for DevWorkspace metadata configmap to be ready",
+		}
+	}
+
+	vol := getVolumeFromConfigMap(cm)
+	podAdditions.Volumes = append(podAdditions.Volumes, *vol)
+	vm := getVolumeMountFromVolume(vol)
+	podAdditions.VolumeMounts = append(podAdditions.VolumeMounts, *vm)
+
+	for idx := range podAdditions.Containers {
+		podAdditions.Containers[idx].Env = append(podAdditions.Containers[idx].Env, getWorkspaceMetaEnvVar()...)
+	}
+
+	for idx := range podAdditions.InitContainers {
+		podAdditions.InitContainers[idx].Env = append(podAdditions.InitContainers[idx].Env, getWorkspaceMetaEnvVar()...)
+	}
+
+	return nil
+}
+
+func syncConfigMapToCluster(specCM *corev1.ConfigMap, api *provision.ClusterAPI) (inSync bool, err error) {
+	clusterCM := &corev1.ConfigMap{}
+	err = api.Client.Get(context.TODO(), types.NamespacedName{Name: specCM.Name, Namespace: specCM.Namespace}, clusterCM)
+
+	switch {
+	case err == nil:
+		if maputils.Equal(specCM.Data, clusterCM.Data) {
+			return true, nil
+		}
+		clusterCM.Data = specCM.Data
+		err = api.Client.Update(context.TODO(), clusterCM)
+	case k8sErrors.IsNotFound(err):
+		err = api.Client.Create(context.TODO(), specCM)
+	default:
+		return false, err
+	}
+	if k8sErrors.IsConflict(err) || k8sErrors.IsAlreadyExists(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+func getSpecMetadataConfigMap(original, flattened *dw.DevWorkspace) (*corev1.ConfigMap, error) {
+	originalYaml, err := yaml.Marshal(original.Spec.Template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal original DevWorkspace yaml: %w", err)
+	}
+
+	flattenedYaml, err := yaml.Marshal(flattened.Spec.Template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal flattened DevWorkspace yaml: %w", err)
+	}
+
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-metadata", original.Status.WorkspaceId), // TODO
```

Comment:
What does this TODO mean and when you plan to address it?

Reply:
I knew I had left one in there somewhere. I'll fix.

Reply:
Done -- moved name generation to naming.go